### DISCUSSION TITLE
Question about "Required, can be None" parameter

### DISCUSSION THREAD

**Discussion by bowenerchen** [fastapi/fastapi#12418] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/items/")
async def read_items(q: Annotated[str | None, Query(min_length=3)] = ...):
    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    return results
```

### Description

https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#required-can-be-none
![QQ_1728534861112](https://github.com/user-attachments/assets/95e967aa-ea9f-46c1-80c8-f13dc7c52e6c)
how to pass a None value to the required query parameter q?

### Operating System

macOS

### Operating System Details

Sonoma 14.7 M2 Max

### FastAPI Version

0.115.0

### Pydantic Version

2.5.3

### Python Version

Python 3.12.4

**Comment by deight93** [comment]
Hello!

To pass a `None` value to a required query parameter q in FastAPI, you can define the parameter as optional with a default value of `None`. This allows you to omit the parameter in the request, passing `None`.

```
from typing import Annotated

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/items/")
async def read_items(q: Annotated[str | None, Query(min_length=3)] = None):
    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    
    if q is None:
        results.update({"q": None})
    
    return results
```

Reference:
https://fastapi.tiangolo.com/tutorial/query-params/#optional-parameters

**Reply by bowenerchen** [reply]
Yep, your codes run properly!
but Is there a problem with this example in the document?
https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#required-can-be-none

**Reply by devlumba** [reply]
I think it kinda undermines the point? It stops being required when being required is the premise?

**Comment by PhysicallyActive** [comment]
This is not really a FastAPI nor a Pydantic issue.

Query parameters are always treated as strings, i.e., everything you pass as a query param will end up being parsed as a string. How do you parse some value to None?
* "" could be an empty string
* "null" or "None" could be the string value

If you really need the nullable parameter to be optional, what you could do is parse the string value via a pydantic model and just "decide" that one specific value represents `None`.

```python
from enum import StrEnum
from typing import Annotated

import pydantic
from fastapi import Depends, FastAPI, Query

app = FastAPI()


class NullString(pydantic.BaseModel):
    q: str | None

    @pydantic.field_validator("q", mode="before")
    @classmethod
    def null_string_to_none(cls, param: str) -> str | None:
        return None if param == "null" else param


async def get_null_string(q: Annotated[str | None, Query(min_length=3)] = ...):
    return NullString(q=q)


@app.get("/items/")
async def read_items(q: Annotated[NullString, Depends(get_null_string)] = ...):
    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    return results
```

**Reply by PhysicallyActive** [reply]
I would not recommend this though. It's a strange way to treat query data and I would assume that refactoring your code would solve your problem rather than finding a solution to this niche problem. Good luck :)

**Reply by bowenerchen** [reply]
Sure, I totally understand what you mean, thank you for your answer. 
However, I saw this usage in the official documentation of FastAPI, but when I actually tested it myself, I found that it could not achieve the expected effect, so I raise this question.

**Reply by sinisaos** [reply]
> Query parameters are always treated as strings, i.e., everything you pass as a query param will end up being parsed as a string. How do you parse some value to None?
> 
> * "" could be an empty string
> * "null" or "None" could be the string value

@PhysicallyActive I think this is a great explanation and I totally agree, but the example raises a `ValueError: [TypeError("'ellipsis' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]`
in the Swagger UI. I think that example should be something like this
```python
from typing import Annotated

from fastapi import FastAPI, Query
from pydantic import BaseModel, Field, field_validator

app = FastAPI()


class NullString(BaseModel):
    q: str | None = Field(None, min_length=3)

    @field_validator("q", mode="before")
    def null_string_to_none(param: str) -> str | None:
        if param == "":
            return "null"
        return param


@app.get("/items/")
async def read_items(q: Annotated[NullString, Query()]):
    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    return results
```
Result is (work correctly both with `curl` or in `Swagger UI`)

```bash
> curl -X 'GET' 'http://localhost:8000/items/?q=' -H 'accept: application/json'
{"items":[{"item_id":"Foo"},{"item_id":"Bar"}],"q":{"q":"null"}} 
> curl -X 'GET' 'http://localhost:8000/items/?q=AA' -H 'accept: application/json'
{"detail":[{"type":"string_too_short","loc":["query","q"],"msg":"String should have at least 3 characters","input":"AA","ctx"{"min_length":3}}]}
> curl -X 'GET' 'http://localhost:8000/items/?q=AAA' -H 'accept: application/json'
{"items":[{"item_id":"Foo"},{"item_id":"Bar"}],"q":{"q":"AAA"}}
```

**Reply by bowenerchen** [reply]
@sinisaos hi, the ellipsis problem, this may be helpful to you: https://github.com/fastapi/fastapi/issues/12313

**Comment by bowenerchen** [comment]
@Kludex Hello, Please take a look is this problem also a documentation issue?

**Reply by Kludex** [reply, marked as answer]
I've converted this to an issue. Thanks.

https://github.com/fastapi/fastapi/issues/12419

**Comment by devlumba** [comment]
been poking this for like 30 minutes, not sure why it's still included in the official docs. Like, you can't possibly pass the null there via query unless it's a string, but then it's just a string. And if you omit, you get an error because the parameter is required.
